File: Products-of-Array-Except-Self.js
//**
// Products of Array Except Self
// Given an integer array nums, return an array output where output[i] is the product of all the elements of nums except nums[i].

// Each product is guaranteed to fit in a 32-bit integer.

// Follow-up: Could you solve it in O(n)
// O(n) time without using the division operation?
//
// **//

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
        productExceptSelf(nums) {

            var product = 1;
            var countZero = 0;
            var productArray = []

            for(let i=0; i<nums.length; i++) {
                if(nums[i]!==0) {
                    product = product*nums[i]
                }
                else {
                    countZero+=1;
                    continue;
                }
            }

            for(let i=0; i<nums.length; i++) {

                if(countZero==0) { // No zero in array
                    productArray.push(product/nums[i])
                }
                else if(countZero>0) { //There is atleast one zero in array

                    //When there is only zero in nums array
                    if(countZero==1 && nums[i]!=0) {
                        productArray.push(0)
                    }
                    else if(countZero==1 && nums[i]==0) {
                        productArray.push(product)
                    }
                    //When there is only zero in nums array
                    if(countZero>1) {
                        productArray.push(0)
                    }
                }
            }

            return productArray;
        }
}

let inst = new Solution();

inst.productExceptSelf([-1,0,1,2,3]);

File: Python/Array_Hashing/Products_of_Array_Except _Self.py
'''
Products of Array Except Self
Solved 
Given an integer array nums, return an array output where output[i] is the product of all the elements of nums except nums[i].

Each product is guaranteed to fit in a 32-bit integer.

Follow-up: Could you solve it in 
O(n) time without using the division operation?

Input: nums = [1,2,4,6]

Output: [48,24,12,8]
'''

from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ln = len(nums)
        sf=[]*ln
        pf=[]
        psf=1
        ppf=1
        fp=[]

        for i, n in enumerate(nums):
            pf.append(ppf)
            ppf = ppf*n   

        for i, n in enumerate(nums[::-1]):
            sf.insert(0, psf)
            psf = psf*n

        for i in range(ln):
            nums[i].append(pf[i]*sf[i])

        return nums



solution = Solution()

strs = [-1,0,1,2,3]

test1 = solution.productExceptSelf(strs)

print(test1)

